cf-editor-form-field-dropdown-sample/src/aem-cf-editor-1/actions/utils.js

```diff
@@ -0,0 +1,168 @@
+ *
+ * @returns {string} if the return value is not null, then it holds an error message describing the missing inputs.
+ *
+ */
+function checkMissingRequestInputs (params, requiredParams = [], requiredHeaders = []) {
+  let errorMessage = null
+
+  // input headers are always lowercase
+  requiredHeaders = requiredHeaders.map(h => h.toLowerCase())
+  // check for missing headers
+  const missingHeaders = getMissingKeys(params.__ow_headers || {}, requiredHeaders)
+  if (missingHeaders.length > 0) {
+    errorMessage = `missing header(s) '${missingHeaders}'`
+  }
+
+  // check for missing parameters
+  const missingParams = getMissingKeys(params, requiredParams)
+  if (missingParams.length > 0) {
+    if (errorMessage) {
+      errorMessage += ' and '
+    } else {
+      errorMessage = ''
+    }
+    errorMessage += `missing parameter(s) '${missingParams}'`
+  }
+
+  return errorMessage
+}
+
+/**
+ *
+ * Extracts the bearer token string from the Authorization header in the request parameters.
+ *
+ * @param {object} params action input parameters.
+ *
+ * @returns {string|undefined} the token string or undefined if not set in request headers.
+ *
+ */
+function getBearerToken (params) {
+  if (params.__ow_headers &&
+      params.__ow_headers.authorization &&
+      params.__ow_headers.authorization.startsWith('Bearer ')) {
+    return params.__ow_headers.authorization.substring('Bearer '.length)
+  }
+  return undefined
+}
+/**
+ *
+ * Returns an error response object and attempts to log.info the status code and error message
+ *
+ * @param {number} statusCode the error status code.
+ *        e.g. 400
+ * @param {string} message the error message.
+ *        e.g. 'missing xyz parameter'
+ * @param {*} [logger] an optional logger instance object with an `info` method
+ *        e.g. `new require('@adobe/aio-sdk').Core.Logger('name')`
+ *
+ * @returns {object} the error object, ready to be returned from the action main's function.
+ *
+ */
+function errorResponse (statusCode, message, logger) {
+  if (logger && typeof logger.info === 'function') {
+    logger.info(`${statusCode}: ${message}`)
+  }
+  return {
+    error: {
+      statusCode,
+      body: {
+        error: message
+      }
+    }
+  }
+}
+
+async function getAemHeaders(params) {
+  const aemHost = params.aemHost;
+  const headers = new Headers();
+  const imsToken = getBearerToken(params);
+
+  const csrfResponse = await fetch(`${aemHost}/libs/granite/csrf/token.json`, {
+    'headers': {
+      'accept': '*/*',
+      "authorization": "Bearer " + imsToken,
+      'X-Api-Key': 'exc_app',
+      'X-Gw-Ims-Org-Id': params.imsOrg
+    },
+    'method': 'GET',
+  });
+  const csrf = await csrfResponse.json();
+
+  headers.set("x-gw-ims-org-id", params.imsOrg);
+  headers.set("x-api-key", 'aem-headless-cf-admin');
+  headers.set("CSRF-Token", csrf.token);
+  headers.set("x-aem-affinity-type", "api");
+  headers.append("Authorization", "Bearer " + imsToken);
+
+  return headers;
+}
```

Comment:
We should consider adding this to our template as this is part of almost all our queries.